calico_kubernetes/policy.py

```diff
@@ -0,0 +1,343 @@
+# Copyright 2015 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+import logging
+from netaddr import IPNetwork, AddrFormatError
+from pycalico.datastore_datatypes import Rule
+
+_log = logging.getLogger(__name__)
+
+
+class PolicyParser(object):
+    """
+    Class which parses Kubernetes annotation-based policy strings and
+    converts them to libcalico Rules objects.
+
+    Usage:
+      parser = PolicyParser(namespace)
+      rule = parser.parse_line("allow tcp from label stage=prod")
+    """
+    def __init__(self, namespace):
+        self.namespace = namespace
+
+    def parse_line(self, policy):
+        """
+        Takes a single line of policy as defined in the annotations of a
+        pod and returns the equivalent libcalico Rule object.
+        :param policy: Policy string from annotations.
+        :return: A Rule object which represent the given policy.
+        """
+        _log.info("Parsing policy line: '%s'", policy)
+        splits = self.validate_policy(policy)
+    
+        # Parse based on the protocol.
+        if "tcp" in splits:
```

Comment:
Yep you're right, was probably a bit silly of me to jump the gun on this one..  When I got started I thought my approach was going to be a bit cheaper than it turned out to be...

I'll take a look at library options here (inc Pyparsing and docopt).  My first impression is that docopt is not the right tool to use here, and I'd rather use something homegrown, but Pyparsing looks interesting and it might be worth swapping over to it.